File: tests/core/test_dnps.cpp
#include <torch/torch.h>
#include <gtest/gtest.h>
#include "dnps.h"

// Test fixture for DNPS tests
class DNPSTest : public ::testing::Test {
protected:
    void SetUp() override {
        // Initialize common parameters
        dim = 16;
        num_experts = 8;
        num_selected = 2;
        batch_size = 2;
        seq_len = 4;
        
        // Create a DNPS instance
        dnps = std::make_unique<sparsemoe::DynamicNeuralPathway>(
            dim, num_experts, num_selected, 1, "sigmoid", 1.0
        );
        
        // Create sample input tensor
        input = torch::randn({batch_size, seq_len, dim});
        
        // Create sample expert weights
        expert_weights.resize(num_experts);
        for (int i = 0; i < num_experts; ++i) {
            expert_weights[i] = torch::randn({dim, dim * 4});
        }
    }

    // Test parameters
    int dim;
    int num_experts;
    int num_selected;
    int batch_size;
    int seq_len;
    
    // Test objects
    std::unique_ptr<sparsemoe::DynamicNeuralPathway> dnps;
    torch::Tensor input;
    std::vector<torch::Tensor> expert_weights;
};

// Test DNPS initialization
TEST_F(DNPSTest, Initialization) {
    // Create DNPS with different parameters
    auto dnps_custom = sparsemoe::DynamicNeuralPathway(
        32, 16, 4, 2, "softmax", 2.0
    );
    
    // Nothing to assert directly as we can't access private members
    // Just test that creation doesn't throw an exception
}

// Test forward pass
TEST_F(DNPSTest, ForwardShape) {
    // Compute routing logits
    torch::Tensor routing_logits = torch::randn({batch_size, seq_len, num_experts});
    
    // Run forward pass
    torch::Tensor output = dnps->forward(input, expert_weights, routing_logits);
    
    // Check output shape
    ASSERT_EQ(output.sizes(), input.sizes());
}

// Test with hierarchical routing
TEST_F(DNPSTest, HierarchicalRouting) {
    // Create DNPS with hierarchical routing
    auto dnps_hierarchical = sparsemoe::DynamicNeuralPathway(
        dim, num_experts, num_selected, 2, "sigmoid", 1.0
    );
    
    // Compute routing logits
    torch::Tensor routing_logits = torch::randn({batch_size, seq_len, num_experts});
    
    // Run forward pass
    torch::Tensor output = dnps_hierarchical.forward(input, expert_weights, routing_logits);
    
    // Check output shape
    ASSERT_EQ(output.sizes(), input.sizes());
}

// Test with different scoring functions
TEST_F(DNPSTest, ScoringFunctions) {
    // Create DNPS with softmax scoring
    auto dnps_softmax = sparsemoe::DynamicNeuralPathway(
        dim, num_experts, num_selected, 1, "softmax", 1.0
    );
    
    // Compute routing logits
    torch::Tensor routing_logits = torch::randn({batch_size, seq_len, num_experts});
    
    // Run forward pass with sigmoid scoring
    torch::Tensor output_sigmoid = dnps->forward(input, expert_weights, routing_logits);
    
    // Run forward pass with softmax scoring
    torch::Tensor output_softmax = dnps_softmax.forward(input, expert_weights, routing_logits);
    
    // Both should produce outputs of the correct shape
    ASSERT_EQ(output_sigmoid.sizes(), input.sizes());
    ASSERT_EQ(output_softmax.sizes(), input.sizes());
}

// Test CUDA support if available
TEST_F(DNPSTest, CudaSupport) {
    if (!torch::cuda::is_available()) {
        GTEST_SKIP() << "CUDA not available";
    }
    
    // Move input and expert weights to CUDA
    torch::Tensor input_cuda = input.to(torch::kCUDA);
    std::vector<torch::Tensor> expert_weights_cuda;
    expert_weights_cuda.resize(num_experts);
    for (int i = 0; i < num_experts; ++i) {
        expert_weights_cuda[i] = expert_weights[i].to(torch::kCUDA);
    }
    
    // Compute routing logits on CUDA
    torch::Tensor routing_logits_cuda = torch::randn(
        {batch_size, seq_len, num_experts}, 
        torch::TensorOptions().device(torch::kCUDA)
    );
    
    // Enable CUDA kernels
    dnps->set_use_cuda_kernels(true);
    
    // Run forward pass on CUDA
    torch::Tensor output_cuda = dnps->forward(input_cuda, expert_weights_cuda, routing_logits_cuda);
    
    // Check output is on CUDA and has correct shape
    ASSERT_TRUE(output_cuda.device().is_cuda());
    ASSERT_EQ(output_cuda.sizes(), input_cuda.sizes());
}

int main(int argc, char **argv) {
    ::testing::InitGoogleTest(&argc, argv);
    return RUN_ALL_TESTS();
}

File: tests/py/test_dnps.py
import unittest
import os
import sys
import torch

# Add project root to path for imports
sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), '../..')))

from sparsemoe.dnps import DNPSLayer

class TestDNPSLayer(unittest.TestCase):
    def setUp(self):
        # Set up common test parameters
        self.batch_size = 2
        self.seq_len = 4
        self.dim = 16
        self.num_experts = 8
        self.num_selected = 2
        
        # Create a sample input tensor
        self.input_tensor = torch.randn(self.batch_size, self.seq_len, self.dim)
        
        # Create DNPSLayer with Python-only implementation
        self.dnps_layer = DNPSLayer(
            dim=self.dim,
            num_experts=self.num_experts,
            num_selected=self.num_selected,
            use_cpp_backend=False  # Force Python implementation
        )
        
    def test_initialization(self):
        """Test that the DNPS layer is initialized correctly."""
        self.assertEqual(self.dnps_layer.dim, self.dim)
        self.assertEqual(self.dnps_layer.num_experts, self.num_experts)
        self.assertEqual(self.dnps_layer.num_selected, self.num_selected)
        
        # Check that all experts were created
        self.assertEqual(len(self.dnps_layer.experts), self.num_experts)
        
        # Check that the router was created
        self.assertIsNotNone(self.dnps_layer.router)
        self.assertEqual(self.dnps_layer.router.in_features, self.dim)
        self.assertEqual(self.dnps_layer.router.out_features, self.num_experts)
        
    def test_forward_shape(self):
        """Test that the forward pass produces output with the correct shape."""
        output = self.dnps_layer(self.input_tensor)
        
        # Output should have the same shape as input
        self.assertEqual(output.shape, self.input_tensor.shape)
        
    def test_forward_different_batch_sizes(self):
        """Test that the forward pass works with different batch sizes."""
        # Test with batch size 1
        input_tensor_1 = torch.randn(1, self.seq_len, self.dim)
        output_1 = self.dnps_layer(input_tensor_1)
        self.assertEqual(output_1.shape, input_tensor_1.shape)
        
        # Test with batch size 4
        input_tensor_4 = torch.randn(4, self.seq_len, self.dim)
        output_4 = self.dnps_layer(input_tensor_4)
        self.assertEqual(output_4.shape, input_tensor_4.shape)
        
    def test_expert_selection(self):
        """Test that routing selects the correct number of experts."""
        # Set up custom simple router to control expert selection behavior
        with torch.no_grad():
            # Set router weights to route to specific experts predictably
            self.dnps_layer.router.weight.zero_()
            self.dnps_layer.router.bias.zero_()
            
            # Make first token always select experts 0,1 and second token select experts 2,3
            self.dnps_layer.router.bias[0] = 10.0
            self.dnps_layer.router.bias[1] = 9.0
            self.dnps_layer.router.bias[2] = 8.0
            self.dnps_layer.router.bias[3] = 7.0
        
        # Compute routing logits manually
        routing_logits = self.dnps_layer.router(self.input_tensor)
        
        # Check that top logits correspond to the expected experts
        top_logits, top_indices = torch.topk(routing_logits, self.num_selected, dim=-1)
        
        # Each token should select the same experts due to our bias settings
        for i in range(self.batch_size):
            for j in range(self.seq_len):
                selected_experts = top_indices[i, j].tolist()
                # Should select experts 0 and 1 based on our bias settings
                self.assertIn(0, selected_experts)
                self.assertIn(1, selected_experts)
    
    def test_hierarchical_routing(self):
        """Test hierarchical routing with expert groups."""
        # Create DNPSLayer with hierarchical routing
        dnps_layer_hierarchical = DNPSLayer(
            dim=self.dim,
            num_experts=8,
            num_selected=2,
            group_size=2,  # 4 groups of 2 experts each
            use_cpp_backend=False  # Force Python implementation
        )
        
        # Forward pass should still work
        output = dnps_layer_hierarchical(self.input_tensor)
        self.assertEqual(output.shape, self.input_tensor.shape)
        
    @unittest.skipIf(not torch.cuda.is_available(), "CUDA not available")
    def test_cuda_support(self):
        """Test that the DNPS layer works on CUDA."""
        # Move model and inputs to CUDA
        dnps_layer_cuda = DNPSLayer(
            dim=self.dim,
            num_experts=self.num_experts,
            num_selected=self.num_selected,
            use_cpp_backend=False
        ).cuda()
        
        input_cuda = self.input_tensor.cuda()
        
        # Forward pass on CUDA
        output_cuda = dnps_layer_cuda(input_cuda)
        
        # Check output is on CUDA and has correct shape
        self.assertTrue(output_cuda.is_cuda)
        self.assertEqual(output_cuda.shape, input_cuda.shape)

if __name__ == '__main__':
    unittest.main()